lib/CL/devices/cuda/pocl-ptx-gen.cc

```diff
@@ -0,0 +1,691 @@
+  std::string err;
+  llvm::raw_string_ostream errs(err);
+  if (llvm::verifyModule(*module->get(), &errs)) {
+    printf("\n%s\n", err.c_str());
+    POCL_ABORT("[CUDA] ptx-gen: module verification failed\n");
+  }
+
+  llvm::StringRef triple =
+      (sizeof(void *) == 8) ? "nvptx64-nvidia-cuda" : "nvptx-nvidia-cuda";
+
+  // Get NVPTX target
+  std::string error;
+  const llvm::Target *target =
+      llvm::TargetRegistry::lookupTarget(triple, error);
+  if (!target) {
+    POCL_MSG_ERR("[CUDA] ptx-gen: failed to get target\n");
+    POCL_MSG_ERR("%s\n", error.c_str());
+    return 1;
+  }
+
+  // TODO: set options
+  llvm::TargetOptions options;
+
+  // TODO: CPU and features?
+  std::unique_ptr<llvm::TargetMachine> machine(target->createTargetMachine(
+      triple, gpu_arch, "+ptx40", options, llvm::None));
+
+  llvm::legacy::PassManager passes;
+
+  // Add pass to emit PTX
+  llvm::SmallVector<char, 4096> data;
+  llvm::raw_svector_ostream ptx_stream(data);
+  if (machine->addPassesToEmitFile(passes, ptx_stream,
+                                   llvm::TargetMachine::CGFT_AssemblyFile)) {
+    POCL_MSG_ERR("[CUDA] ptx-gen: failed to add passes\n");
+    return 1;
+  }
+
+  // Run passes
+  passes.run(**module);
+
+  std::string ptx = ptx_stream.str(); // flush
+  return pocl_write_file(ptx_filename, ptx.c_str(), ptx.size(), 0, 0);
+}
+
+void pocl_add_kernel_annotations(llvm::Module *module, const char *kernel) {
+  llvm::LLVMContext &context = module->getContext();
+
+  // Remove existing nvvm.annotations metadata since it is sometimes corrupt
+  auto nvvm_annotations = module->getNamedMetadata("nvvm.annotations");
+  if (nvvm_annotations)
+    nvvm_annotations->eraseFromParent();
+
+  // Add nvvm.annotations metadata to mark kernel entry point
+  nvvm_annotations = module->getOrInsertNamedMetadata("nvvm.annotations");
+
+  // Get handle to function
+  auto func = module->getFunction(kernel);
+  if (!func)
+    POCL_ABORT("[CUDA] ptx-gen: kernel function not found in module\n");
+
+  // Create metadata
+  llvm::Constant *one =
+      llvm::ConstantInt::getSigned(llvm::Type::getInt32Ty(context), 1);
+  llvm::Metadata *md_f = llvm::ValueAsMetadata::get(func);
+  llvm::Metadata *md_n = llvm::MDString::get(context, "kernel");
+  llvm::Metadata *md_1 = llvm::ConstantAsMetadata::get(one);
+
+  std::vector<llvm::Metadata *> v_md = {md_f, md_n, md_1};
+  llvm::MDNode *node = llvm::MDNode::get(context, v_md);
+  nvvm_annotations->addOperand(node);
+}
+
+void pocl_erase_function_and_callers(llvm::Function *func) {
+  if (!func)
+    return;
+
+  std::vector<llvm::Value *> callers(func->users().begin(),
+                                     func->users().end());
+  for (auto U = callers.begin(); U != callers.end(); U++) {
+    llvm::CallInst *call = llvm::dyn_cast<llvm::CallInst>(*U);
+    if (!call)
+      continue;
+    call->eraseFromParent();
+  }
+  func->eraseFromParent();
+}
+
+void pocl_cuda_fix_printf(llvm::Module *module) {
```

Comment:
This is an interesting function. Can you add a brief comment what goes on here? CUDA doesn't have varargs support ?